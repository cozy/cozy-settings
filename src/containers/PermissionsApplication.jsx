import React from 'react'

import { APPS_DOCTYPE } from 'doctypes'
import { Q, useQuery, isQueryLoading, hasQueryBeenLoaded } from 'cozy-client'
import Spinner from 'cozy-ui/transpiled/react/Spinner'
import Typography from 'cozy-ui/transpiled/react/Typography'
import { useI18n } from 'cozy-ui/transpiled/react'
import Page from 'components/Page'
import PageTitle from 'components/PageTitle'
import { withRouter } from 'react-router-dom'

const PermissionsApplication = ({ match }) => {
  const { t } = useI18n()
  const appName = match.params.app

  const queryResult = useQuery(
    Q(APPS_DOCTYPE).getById('io.cozy.apps/' + appName),
    {
      as: 'io.cozy.apps/' + appName
    }
  )
  console.log(queryResult)

  return (
    <Page narrow>
      {isQueryLoading(queryResult) && !hasQueryBeenLoaded(queryResult) ? (
        <Spinner size="large" className="u-flex u-flex-justify-center u-mt-1" />
      ) : queryResult.fetchStatus === 'failed' ? (
        <Typography variant="body1" className="u-mb-1-half">
          {t('Permissions.failedRequest')}
        </Typography>
      ) : (
        <div>
          <PageTitle>{appName.toUpperCase()}</PageTitle>
          {Object.entries(queryResult.data[0].attributes.permissions).map(
            ([key, value]) => (
              <React.Fragment key={key}>
                <Typography variant="h4">
                  {value.type} : {value.verbs ? value.verbs.join(' / ') : 'ALL'}
                </Typography>
              </React.Fragment>
            )
          )}
        </div>
      )}
    </Page>
  )
}

export default withRouter(PermissionsApplication)
