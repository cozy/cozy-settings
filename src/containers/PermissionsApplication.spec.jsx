import { render } from '@testing-library/react'
import React from 'react'
import PermissionsApplication from './PermissionsApplication'
import { Q, useQuery, isQueryLoading, hasQueryBeenLoaded } from 'cozy-client'

jest.mock('cozy-ui/transpiled/react', () => {
  return { useI18n: () => ({ t: x => x }) }
})

jest.mock('react-router-dom', () => {
  return {
    withRouter: Component => {
      const match = { params: { app: 'Drive' } }
      // eslint-disable-next-line react/display-name
      return () => <Component match={match} />
    },
    Link: () => {
      // eslint-disable-next-line react/display-name
      return ({ to }) => <div data-testid="Link">{to}</div>
    }
  }
})

jest.mock('cozy-client')

jest.mock('components/Page', () => {
  // eslint-disable-next-line react/display-name
  return ({ narrow, children }) => (
    <div data-testid="page" data-narrow={narrow}>
      {children}
    </div>
  )
})

jest.mock('components/PageTitle', () => {
  // eslint-disable-next-line react/display-name
  return ({ children }) => <div data-testid="pageTitle">{children}</div>
})
jest.mock('cozy-ui/transpiled/react/Spinner', () => {
  // eslint-disable-next-line react/display-name
  return ({ size }) => <div data-testid="Spinner" data-size={size}></div>
})

describe('PermissionsApplication', () => {
  beforeEach(() => {
    const queryResult = {
      fetchStatus: 'loaded',
      data: [
        {
          attributes: {
            permissions: {
              files: {
                type: 'io.cozy.files',
                description: 'Required to access the files'
              },
              allFiles: {
                type: 'io.cozy.files.*',
                description: 'Required to access the files'
              },
              apps: {
                type: 'io.cozy.apps',
                description:
                  'Required by the cozy-bar to display the icons of the apps',
                verbs: ['GET']
              }
            }
          }
        }
      ]
    }
    isQueryLoading.mockReturnValue(true)
    hasQueryBeenLoaded.mockReturnValue(true)
    Q.mockReturnValue({ getById: () => 'kfrf' })
    useQuery.mockReturnValue(queryResult)
  })
  it('should display appName when query has been loaded', () => {
    hasQueryBeenLoaded.mockReturnValue(true)
    const { container } = render(<PermissionsApplication />)
    expect(container).toMatchSnapshot()
  })

  it('should render a spinner when query is loading and has not been loaded', () => {
    isQueryLoading.mockReturnValue(true)
    hasQueryBeenLoaded.mockReturnValue(false)
    const { queryByTestId } = render(<PermissionsApplication />)
    expect(queryByTestId('Spinner')).toBeTruthy()
  })

  it('should display appName when query is not loading', () => {
    isQueryLoading.mockReturnValue(false)
    const { queryByText } = render(<PermissionsApplication />)
    expect(queryByText('DRIVE')).toBeTruthy()
  })

  it('should render Permissions.failedRequest when query status is failed', () => {
    useQuery.mockReturnValue({ fetchStatus: 'failed' })
    const { queryByText } = render(<PermissionsApplication />)
    expect(queryByText('Permissions.failedRequest')).toBeTruthy()
  })
})
