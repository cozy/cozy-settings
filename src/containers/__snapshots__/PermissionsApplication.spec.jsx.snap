// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PermissionsApplication should display appName when query has been loaded 1`] = `
<div>
  <div
    data-narrow="true"
    data-testid="page"
  >
    <div>
      <div
        data-testid="pageTitle"
      >
        DRIVE
      </div>
    </div>
  </div>
</div>
`;
